dial/pipeline.go

```diff
@@ -0,0 +1,313 @@
+
+// Complete fills in the result of the dial request. It can only be invoked once per Request.
+// After the first completion, further calls to complete will fail with error.
+//
+// Complete invokes all callbacks associated with this Request, and closes the notifyCh.
+func (req *Request) Complete(conn inet.Conn, err error) error {
+	if !atomic.CompareAndSwapInt32(&req.status, requestInflight, requestFinished) {
+		return ErrDialRequestInvalidStatus
+	}
+
+	req.conn, req.err = conn, err
+	for i := len(req.callbacks); i > 0; i-- {
+		req.callbacks[i-1]()
+	}
+
+	close(req.notifyCh)
+	return nil
+}
+
+func (req *Request) IsComplete() bool {
+	return atomic.LoadInt32(&req.status) == requestFinished
+}
+
+// Values returns the connection and error fields from this request.
+// Both return values may be nil or incoherent unless the request has completed (see IsComplete).
+func (req *Request) Values() (inet.Conn, error) {
+	return req.conn, req.err
+}
+
+// AddCallback adds a function that will be invoked when this request completes, either in success
+// or in failure.
+func (req *Request) AddCallback(cb func()) {
+	req.callbacks = append(req.callbacks, cb)
+}
+
+// Job represents a dial attempt to a single multiaddr. It is associated 1-* to a Request.
+// It can carry its own context and its own callbacks.
+type Job struct {
+	status int32
+	req    *Request
+
+	ctx  context.Context
+	addr ma.Multiaddr
+
+	callbacks []func()
+
+	// result.
+	tconn tpt.Conn
+	err   error
+
+	// Channel where completed dials should be sent.
+	completeCh chan *Job
+}
+
+func NewDialJob(ctx context.Context, req *Request, addr ma.Multiaddr) *Job {
+	return &Job{req: req, ctx: ctx, addr: addr, status: jobInflight}
+}
+
+func (j *Job) Cancelled() bool {
+	select {
+	case <-j.req.ctx.Done():
+		return true
+	case <-j.ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (j *Job) AddCallback(cb func()) {
+	j.callbacks = append(j.callbacks, cb)
+}
+
+func (j *Job) Complete(conn tpt.Conn, err error) error {
+	if !atomic.CompareAndSwapInt32(&j.status, jobInflight, jobFinished) {
+		return ErrDialJobInvalidStatus
+	}
+
+	j.tconn, j.err = conn, err
+	for i := len(j.callbacks); i > 0; i-- {
+		j.callbacks[i-1]()
+	}
+
+	j.completeCh <- j
+	return nil
+}
+
+type dialJobs []*Job
+
+func (djs *dialJobs) sift() (success dialJobs, failed dialJobs) {
+	for _, dj := range *djs {
+		if dj.err == nil {
+			success = append(success, dj)
+			continue
+		}
+		failed = append(failed, dj)
+	}
+	return success, failed
+}
+
+type AddConnFn func(tc tpt.Conn, dir inet.Direction) (inet.Conn, error)
+
+type Pipeline struct {
+	lk  sync.RWMutex
+	ctx context.Context
+
+	net inet.Network
+
+	reqPreparers []RequestPreparer
+	planner      Planner
+	throttler    Throttler
+	jobPreparers []JobPreparer
+	executor     Executor
+	selector     Selector
+
+	throttleCh chan *Job
+	dialCh     chan *Job
+
+	addConnFn AddConnFn
+}
+
+func (p *Pipeline) Component(name string, comp interface{}) error {
```

Comment:
We *can* do this, or we could just have separate `RegisterPlanner`, etc... functions. What was the motivation?

One potential motivation is that *this* version allows us to use the same component multiple times (e.g., we can have a *single* component implement the preparer and the planner).

Reply:
Absolutely. This choice is simplistic and controversial on purpose to incite dialogue. I haven't landed on a pipeline configuration API I feel good about yet. 

I'm leaning towards providing functions to manage `{Request,Job}Preparers` positionally, e.g. `AddLast, AddFirst, AddAfter, AddBefore, Replace` (hence name argument).

I also think the API should make it possible to change components at runtime (some components embed `io.Closer` to this avail.